python/ray/train/base_trainer.py

```diff
@@ -164,8 +176,169 @@ def __init__(
+            from ray import tune
+
+            experiment_name = "unique_experiment_name"
+            upload_dir = "s3://bucket"
+            experiment_dir = os.path.join(upload_dir, experiment_name)
+
+            # Pretend this is a large object that's been loaded
+            large_data = {}
+            # Use an object reference to share this object across the cluster
+            large_data_ref = ray.put(large_dataset)
+
+            datasets = {"train": ray.data.from_items([{"a": i} for i in range(10)])}
+
+            def train_loop_per_worker(config):
+                pass
+
+            train_loop_config = {"obj_ref": large_data_ref}
+
+            if TorchTrainer.can_restore(experiment_dir):
+                trainer = TorchTrainer.restore(
+                    experiment_dir,
+                    train_loop_per_worker=train_loop_per_worker,
+                    train_loop_config=train_loop_config,
+                    datasets=datasets,
+                )
+            else:
+                trainer = TorchTrainer(
+                    train_loop_per_worker,
+                    train_loop_config,
+                    datasets=datasets,
+                    run_config=air.RunConfig(
+                        name=experiment_name,
+                        sync_config=tune.SyncConfig(upload_dir=upload_dir),
+                        # Tip: Add trial-level fault-tolerance on top.
+                        failure_config=air.FailureConfig(max_failures=3),
+                    ),
+                )
+
+            result = trainer.fit()
+
+
+        Args:
+            path: The path to the experiment directory of the training run to restore.
+                This can be a local path or a remote URI if the experiment was
+                uploaded to the cloud.
+            datasets: Re-specified datasets used in the original training run.
+                This must include all the datasets that were passed in the
+                original trainer constructor.
+            preprocessor: Optionally re-specified preprocessor that was passed in
+                the original trainer constructor. This should be used to re-supply
+                the preprocessor if it is not restorable in a new Ray cluster.
+                This preprocessor will be fit at the start before resuming training.
+                If no preprocessor is passed in restore, then the old preprocessor
+                will be loaded from the latest checkpoint and will not be re-fit.
+            scaling_config: Optionally re-specified scaling config. This can be
+                modified to be different from the original spec.
+            **kwargs: Other optionally re-specified arguments, passed in by subclasses.
+
+        Raises:
+            ValueError: If all datasets were not re-supplied on restore.
+
+        Returns:
+            BaseTrainer: A restored instance of the class that is calling this method.
+        """
+        assert cls.can_restore(path), (
```

Comment:
Can we avoid using assertions and instead raise Value/Runtime errors?

Reply:
Got it, asserts should only be used for internal sanity checks. Not external validation.